src/lora_rxdelay.erl

```diff
@@ -0,0 +1,232 @@
+    Requested = maps:get(rx_delay, ApiSettings, Actual),
+    case Requested == Actual of
+        true ->
+            DeviceState#{rx_delay_state => ?RX_DELAY_ESTABLISHED};
+        false when Requested =< 15 ->
+            %% Track requested value for new `rx_delay` without actually changing to it.
+            DeviceState#{rx_delay_state => ?RX_DELAY_CHANGE}
+    end.
+
+-spec adjust_on_join(Device :: router_device:device()) -> DeviceState :: map().
+adjust_on_join(Device) ->
+    %% LoRaWAN Join-Accept implies `rx_timing_setup_ans` because
+    %% `RXDelay` was in the Join-Accept.  A subsequent uplink implies
+    %% that the device acknowledged the Join-Accept and thus, RXDelay.
+    {_, DeviceState, _} = adjust(Device, [rx_timing_setup_ans], []),
+    DeviceState.
+
+-spec adjust(DeviceState0 :: map(), UplinkFOpts :: list(), FOpts0 :: list()) ->
+    {RxDelay :: non_neg_integer(), DeviceState1 :: map(), FOpts1 :: list()}.
+adjust(DeviceState0, UplinkFOpts, FOpts0) ->
+    %% When state is undefined, it's probably a test that omits device_update()
+    %% or didn't wait long enough for the async/cast to complete.
+    State = maps:get(rx_delay_state, DeviceState0, unknown_state),
+    Actual = maps:get(rx_delay_actual, DeviceState0, 0),
+    Answered = lists:member(rx_timing_setup_ans, UplinkFOpts),
+    %% Handle sequential changes to rx_delay; e.g, a previous change
+    %% was ack'd, and now there's potentially another new RxDelay value.
+    {RxDelay, DeviceState1, FOpts1} =
+        case {State, Answered} of
+            %% Entering `RX_DELAY_CHANGE' state occurs in maybe_update().
+            {?RX_DELAY_ESTABLISHED, _} ->
+                {Actual, DeviceState0, FOpts0};
+            {?RX_DELAY_CHANGE, _} ->
+                %% Console changed `rx_delay' value.
+                Requested = maps:get(rx_delay, DeviceState0),
+                FOpts = [{rx_timing_setup_req, Requested} | FOpts0],
+                DeviceState = DeviceState0#{rx_delay_state => ?RX_DELAY_REQUESTED},
+                {Actual, DeviceState, FOpts};
+            {?RX_DELAY_REQUESTED, false} ->
+                %% Router sent downlink request, Device has yet to ACK.
+                Requested = maps:get(rx_delay, DeviceState0),
+                FOpts = [{rx_timing_setup_req, Requested} | FOpts0],
+                {Actual, DeviceState0, FOpts};
+            {?RX_DELAY_REQUESTED, true} ->
+                %% Device responded with ACK, so Router can apply the requested rx_delay.
+                Requested = maps:get(rx_delay, DeviceState0),
+                Map = #{rx_delay_actual => Requested, rx_delay_state => ?RX_DELAY_ESTABLISHED},
+                {Requested, maps:merge(DeviceState0, Map), FOpts0};
+            {unknown_state, _} ->
+                lager:error("rx_delay state=unknown device-id=~p", [router_device:id(DeviceState0)]),
+                {Actual, DeviceState0, FOpts0}
+        end,
+    {RxDelay, DeviceState1, FOpts1}.
+
+-ifdef(TEST).
```

Comment:
This should be EUNIT and not TEST.  The TEST define is intended for CT tests.

Reply:
Fixed in commit ad8b6a6